eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,274 @@
+			_, rOk := currentCallFrame.AccessedSlots.Reads[slotHex]
+			_, wOk := currentCallFrame.AccessedSlots.Writes[slotHex]
+			if !rOk && !wOk {
+				currentCallFrame.AccessedSlots.Reads[slotHex] = append(currentCallFrame.AccessedSlots.Reads[slotHex], t.env.StateDB.GetState(addr, slot).Hex())
+			}
+		} else if opcode == "SSTORE" {
+			incrementCount(currentCallFrame.AccessedSlots.Writes, slotHex)
+		} else if opcode == "TLOAD" {
+			incrementCount(currentCallFrame.AccessedSlots.TransientReads, slotHex)
+		} else if opcode == "TSTORE" {
+			incrementCount(currentCallFrame.AccessedSlots.TransientWrites, slotHex)
+		}
+	}
+}
+
+func (t *callTracerWithOpcodes) storeKeccak(opcode string, scope tracing.OpContext) {
+	if opcode == "KECCAK256" {
+		dataOffset := peepStack(scope.StackData(), 0).Uint64()
+		dataLength := peepStack(scope.StackData(), 1).Uint64()
+		memory := scope.MemoryData()
+		keccak := make([]byte, dataLength)
+		copy(keccak, memory[dataOffset:dataOffset+dataLength])
+		t.Keccak = append(t.Keccak, keccak)
+	}
+}
+
+func (t *callTracerWithOpcodes) detectOutOfGas(gas uint64, cost uint64, opcode string, currentCallFrame callFrameWithOpcodes) {
+	if gas < cost || (opcode == "SSTORE" && gas < 2300) {
+		currentCallFrame.OutOfGas = true
+	}
+}
+
+func (t *callTracerWithOpcodes) handleExtOpcodes(opcode string, currentCallFrame callFrameWithOpcodes, stackTop3 partialStack) {
+	if strings.HasPrefix(opcode, "EXT") {
+		addr := common.HexToAddress(stackTop3[0].Hex())
+		ops := []string{}
+		for _, item := range t.lastThreeOpCodes {
+			ops = append(ops, item)
+		}
+		last3OpcodeStr := strings.Join(ops, ",")
+
+		// only store the last EXTCODE* opcode per address - could even be a boolean for our current use-case
+		// [OP-051]
+		if !strings.Contains(last3OpcodeStr, ",EXTCODESIZE,ISZERO") {
+			currentCallFrame.ExtCodeAccessInfo = append(currentCallFrame.ExtCodeAccessInfo, addr)
+		}
+	}
+}
+func (t *callTracerWithOpcodes) handleAccessedContractSize(opcode string, scope tracing.OpContext, currentCallFrame callFrameWithOpcodes) {
+	// [OP-041]
+	if isEXTorCALL(opcode) {
+		n := 0
+		if !strings.HasPrefix(opcode, "EXT") {
+			n = 1
+		}
+		addr := common.BytesToAddress(peepStack(scope.StackData(), n).Bytes())
+
+		if _, ok := currentCallFrame.ContractSize[addr]; !ok && !isAllowedPrecompile(addr) {
+			currentCallFrame.ContractSize[addr] = len(t.env.StateDB.GetCode(addr))
+		}
+	}
+}
+
+func peepStack(stackData []uint256.Int, n int) *uint256.Int {
+	return &stackData[len(stackData)-n-1]
+}
+
+func isEXTorCALL(opcode string) bool {
+	return strings.HasPrefix(opcode, "EXT") ||
+		opcode == "CALL" ||
+		opcode == "CALLCODE" ||
+		opcode == "DELEGATECALL" ||
+		opcode == "STATICCALL"
+}
+
+// not using 'isPrecompiled' to only allow the ones defined by the ERC-7562 as stateless precompiles
+// [OP-062]
+func isAllowedPrecompile(addr common.Address) bool {
+	addrInt := addr.Big()
+	return addrInt.Cmp(big.NewInt(0)) == 1 && addrInt.Cmp(big.NewInt(10)) == -1
```

Comment:
Add constant list (could bit a bitmap) of precompiles. we already know if one more 0x100, defined in eip-7212